### ISSUE TITLE
asInstanceOf() should retain description

### ISSUE THREAD

**Issue by scordio** [joel-costigliola/assertj-core#1517]
Currently `asInstanceOf()` is not keeping the existing assertion description.

To achieve it, the return statement of:

https://github.com/joel-costigliola/assertj-core/blob/aee33e05bdba3e39376abd9083941c5fc632a905/src/main/java/org/assertj/core/api/AbstractAssert.java#L371-L373

should be changed to:

```java
return instanceOfAssertFactory.createAssert(actual).as(info.description());
```

but this cannot work with the current method declaration because of the type parameter `ASSERT` extending `Assert<?, ?>` instead of `Assert<ASSERT, ?>`.

To fix it, some changes would be needed which cause the replacement of all the `assertThat()` abstract return types with the corresponding concrete ones.

My main question is: why are some `assertThat()` returning an abstract class instead of the concrete one, while others (e.g., `assertThat(Class<?>`) are already returning concrete types? 
Is there a design reason behind it which I'm not understanding?

**Comment by joel-costigliola** [comment]
That was a long time ago, so I'm not sure I completely remember, I think the rationale was to let users to return their own implementation assertion class. Having said that, I'm not too sure if that ever worked and it was done consistently. I'll dig a bit more in the commit history if I can find a better reason.

Back to the initial issue, it is true that description is not propagated but the overridden error message is not either. I think here we need to do call [`withAssertionState`](https://github.com/joel-costigliola/assertj-core/blob/054233d026c08859ce3ee7147ce08ef7a4e654c6/src/main/java/org/assertj/core/api/AbstractAssert.java#L821)

**Comment by scordio** [comment]
I assume you mean something like:

```java
return instanceOfAssertFactory.createAssert(actual).withAssertionState(this); // Not working
```

but it doesn't work at the moment because `createAssert()` returns an `Assert` as generic base while `withAssertionState()` is available only in `AbstractAssert`.
`Assert#as()` and `Assert#withRepresentation()` could partially cover what `withAssertionState()` does, but the call to `overridingErrorMessage()` and the replacement of the internal `objects` are missing.

I think it could make sense to change the generic base of `asInstanceOf()` to `AbstractAssert` to allow the call of `withAssertionState()`, but the abstract/concrete issue would still be there.

I'll wait for your analysis before going ahead. Thanks!

**Comment by joel-costigliola** [comment]
One thing I don't want is to expose `withAssertionState` in `Assert` as it would show up for users when it is only an internal method, this means we need to replace `Assert` by `AbstractAssert` in `InstanceOfAssertFactory`.

Just "playing" a bit with generics, I managed to get the code to compile with:

```java
public class InstanceOfAssertFactory<T, ASSERT extends AbstractAssert<?, ?>> implements AssertFactory<Object, ASSERT>
```
and
```java
  public <ASSERT extends AbstractAssert<?, ?>> ASSERT asInstanceOf(InstanceOfAssertFactory<?, ASSERT> instanceOfAssertFactory) {
    return (ASSERT) instanceOfAssertFactory.createAssert(actual).withAssertionState(myself);
  }
```

A cast to `ASSERT` is need as Java would otherwise complains about `AbstractAssert<?, ?>` (the usual capture error when using `?`)

**Comment by scordio** [comment]
I think I did the same but I had issues in other places where I was using the static factories. At the moment I cannot check, I'll come back on this later during the day. Thanks so far!

**Comment by joel-costigliola** [comment]
No problem, take your time.

**Comment by scordio** [comment]
I made some progress and I'm gonna open a PR to continue the discussion :-)

**Comment by scordio** [comment]
Fixed in a2fe4b281f8c6a9a799e4544be9c72f128c137be.
